SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestHandoverReplication/SpatialTestHandoverActorComponentReplication.cpp

```diff
@@ -0,0 +1,195 @@
+{
+	Super::OnAuthorityGained();
+
+	if (TestStage == EHandoverReplicationTestStage::ChangeValuesToDefaultOnGainingAuthority)
+	{
+		SetTestValues(HandoverReplicationTestValues::BasicTestPropertyValue);
+
+		TestStage = EHandoverReplicationTestStage::Final;
+
+		SetActorLocation(HandoverReplicationTestValues::Server1Position);
+	}
+}
+
+void AHandoverReplicationTestCube::GetLifetimeReplicatedProps(TArray<FLifetimeProperty>& OutLifetimeProps) const
+{
+	Super::GetLifetimeReplicatedProps(OutLifetimeProps);
+
+	DOREPLIFETIME(ThisClass, ReplicatedTestProperty);
+}
+
+/**
+ *	This tests handover values replication on actors and attached components,
+ *	as well as a corner case from UNR-4447: when a server sets a handover property to
+ *	its default value, this property's change isn't passed to other servers.
+ *
+ *	The overall flow is as follows:
+ *		* Spawn a cube on Server 1
+ *		* Modify handover values on the cube and on an attached component
+ *		* Move this cube to Server 2's authority area
+ *		* Change these values to default state
+ *		* Check that the value change was registered
+ */
+
+ASpatialTestHandoverActorComponentReplication::ASpatialTestHandoverActorComponentReplication()
+	: Super()
+{
+	Author = TEXT("Dmitrii Kozlov");
+	Description = TEXT("Test handover replication for an actor and its component");
+
+	bReplicates = true;
+}
+
+void ASpatialTestHandoverActorComponentReplication::GetLifetimeReplicatedProps(TArray<FLifetimeProperty>& OutLifetimeProps) const
+{
+	Super::GetLifetimeReplicatedProps(OutLifetimeProps);
+
+	DOREPLIFETIME(ThisClass, HandoverCube);
+}
+
+void ASpatialTestHandoverActorComponentReplication::PrepareTest()
+{
+	Super::PrepareTest();
+
+	AddStep(TEXT("Check initial settings"), FWorkerDefinition::AllServers, nullptr, [this]() {
+		ASpatialWorldSettings* WorldSettings = Cast<ASpatialWorldSettings>(GetWorld()->GetWorldSettings());
+
+		if (AssertIsValid(WorldSettings, TEXT("World Settings of correct type")))
+		{
+			AssertTrue(IsValid(WorldSettings->GetMultiWorkerSettingsClass())
+						   && WorldSettings->GetMultiWorkerSettingsClass()->IsChildOf<USpatialTestHandoverReplicationMultiWorkerSettings>(),
+					   TEXT("MultiWorkerSettings should be of class USpatialTestHandoverReplicationMultiWorkerSettings"));
+		}
+
+		FinishStep();
+	});
+
+	AddStep(TEXT("Server 1 spawns a HandoverCube"), FWorkerDefinition::Server(1), nullptr, [this]() {
+		HandoverCube = GetWorld()->SpawnActor<AHandoverReplicationTestCube>(HandoverReplicationTestValues::Server1Position,
+																			FRotator::ZeroRotator, FActorSpawnParameters());
+		SaveHandoverCube(HandoverCube);
+		RegisterAutoDestroyActor(HandoverCube);
+		FinishStep();
+	});
+
+	auto AddWaitingStep = [this](const FString& StepName, const FWorkerDefinition& WorkerDefinition, TFunction<void()> TickFunction) {
+		constexpr float StepTimeLimit = 10.0f;
+
+		AddStep(
+			StepName, WorkerDefinition, nullptr, nullptr,
+			[this, TickFunction](float) {
+				TickFunction();
```

Comment:
This is confusing to read. Does it tick only once? Or is it relying on a RequireTrue instruction inside the tick function to prevent FinishStep from doing its job ?

Reply:
The second thought is correct - this is based off of https://github.com/spatialos/UnrealGDK/blob/master/SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestHandoverReplication/SpatialTestHandoverReplication.cpp#L106...L111, just reducing boilerplate somewhat.

Reply:
The only idea that comes to my mind is to add some comments detailing how why I use this as it's a rather convenient feature.

Reply:
@ImprobableNic this is how our polling test functions work as described in the bit I wrote in our testing docs I wrote yesterday, Miron implemented them with RequireTrue : ), the finishStep() bit is confusing I agree but I guess it makes them more in line with a non polling AssertTrue step.

Reply:
This is not too bad, I just wonder about someone using it without realising this is a dependency.